drivers/clock_control/clock_control_mchp_xec.c

```diff
@@ -0,0 +1,640 @@
+ * Core, CPU, and Bus clocks are always on except in deep sleep state.
+ * Peripheral clocks can be gated off if the peripheral's PCR sleep enable
+ * is set and the peripheral indicates it does not need a clock by clearing
+ * its PCR CLOCK_REQ read-only status.
+ * Peripheral slow clock my be turned on by writing a non-zero divider value
+ * to its PCR control register.
+ */
+static int xec_clock_control_on(const struct device *dev,
+				clock_control_subsys_t sub_system)
+{
+	return xec_cc_on(dev, sub_system, true);
+}
+
+/*
+ * Turn off clock source.
+ * Core, CPU, and Bus clocks are always on except in deep sleep when PLL is
+ * turned off. Exception is 32 KHz clock.
+ * Peripheral clocks are gated off when the peripheral's sleep enable is set
+ * and the peripheral indicates is no longer needs a clock by de-asserting
+ * its read-only PCR CLOCK_REQ bit.
+ * Peripheral slow clock can be turned off by writing 0 to its control register.
+ */
+static inline int xec_clock_control_off(const struct device *dev,
+					clock_control_subsys_t sub_system)
+{
+	return xec_cc_on(dev, sub_system, false);
+}
+
+/*
+ * MEC172x clock subsystem:
+ * Two main clock domains: PLL and Peripheral-32K. Each domain's 32 KHz source
+ * can be selected from one of three inputs:
+ *  internal silicon OSC +/- 2% accuracy
+ *  external crystal connected parallel or single ended
+ *  external 32 KHz 50% duty cycle waveform on 32KHZ_IN pin.
+ * PLL domain supplies 96 MHz, 48 MHz, and other high speed clocks to all
+ * peripherals except those in the Peripheral-32K clock domain. The slow clock
+ * is derived from the 48 MHz produced by the PLL.
+ *   ARM Cortex-M4 core input: 96MHz
+ *   AHB clock input: 48 MHz
+ *   Fast AHB peripherals: 96 MHz internal and 48 MHz AHB interface.
+ *   Slow clock peripherals: PWM,  TACH, PROCHOT
+ * Peripheral-32K domain peripherals:
+ *   WDT, RTC, RTOS timer, hibernation timers, week timer
+ *
+ * Peripherals using both PLL and 32K clock domains:
+ *   BBLED, RPMFAN
+ */
+static int xec_clock_control_get_subsys_rate(const struct device *dev,
+					     clock_control_subsys_t sub_system,
+					     uint32_t *rate)
+{
+	struct pcr_regs *const pcr = XEC_PCR_REGS_BASE(dev);
+	uint32_t bus = (uint32_t)sub_system;
+	uint32_t temp = 0;
+
+	switch (bus) {
+	case MCHP_XEC_PCR_CLK_CORE:
+	case MCHP_XEC_PCR_CLK_PERIPH_FAST:
+		*rate = MHZ(96);
+		break;
+	case MCHP_XEC_PCR_CLK_CPU:
+		/* if PCR PROC_CLK_CTRL is 0 the chip is not executing code */
+		*rate = MHZ(96) / pcr->PROC_CLK_CTRL;
+		break;
+	case MCHP_XEC_PCR_CLK_BUS:
+	case MCHP_XEC_PCR_CLK_PERIPH:
+		*rate = MHZ(48);
+		break;
+	case MCHP_XEC_PCR_CLK_PERIPH_SLOW:
+		temp = pcr->SLOW_CLK_CTRL;
+		if (pcr->SLOW_CLK_CTRL) {
+			*rate = MHZ(48) / temp;
+		} else {
+			*rate = 0; /* slow clock off */
+		}
+		break;
+	default:
+		*rate = 0;
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+#if defined(CONFIG_PM)
+void xec_clock_control_system_sleep_enable(bool is_deep)
```

Comment:
@scottwcpg how would this play with Zephyr PM subsystem

Reply:
PCR HW contains the sleep control and this driver owns PCR. Instead of SoC sleep code writing PCR sleep enable directly, it would call xec_clock_control_system_sleep_enable() and xec_clock_control_system_sleep_disable().  These two functions need to be exposed through a public header.

Reply:
@albertofloyd I noticed other EC's have moved most if not all functionality out of soc_init() into drivers.  Is this a goal for Zephyr? Clock control was the closest driver model for the MEC PCR clock features.

Reply:
Got it, will wait to see SoC updates for LPM to understand further this new approach